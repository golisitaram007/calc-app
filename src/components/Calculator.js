import React, { Component } from 'react';
import Result from './Result';
import Keypad from './Keypad';

export default class Calculator extends Component {
    
    state = {
        expression: '',
        result: '',
        prevChar: '',
    }


    updateExpression = e => {
        switch (e) {
            case 'clear':
                this.clearResult('');
                break;
            case 'Backspace':
                this.clearLastExp(this.state.expression);
                break;
            case 'Enter':
                this.calculateResult(this.state.expression);
                break;
            default:
                this.setExpression(e);
        }
    }

    clearResult = val => {
        this.setState({
            expression: val,
            result: val,
            prevChar: val
        })
    }

    updateResult = result => {
        this.setState({
            result
        }, () => {
            this.setState({
                expression: ''
            })
        })
    }

    setExpression = k => {
        let currentExp = this.state.expression;
        const prevChar = this.state.prevChar;
        if('+-*/'.indexOf(prevChar) > -1 && isNaN(k)) {
            // if() {
                currentExp = currentExp.replace(/.$/, k);
                this.setState({
                    expression: currentExp
                })
            // }else {
            //     this.setState({
            //         expression: currentExp + k
            //     })
            // }
        }else {
            this.setState({
                expression: currentExp + k
            })
        }
        this.setState({ prevChar: k })
    }

    clearExpression = val => {
        this.setState({
            expression: val
        })
    }

    clearLastExp = (currentExp) => {
        this.setState({
            expression: currentExp.slice(0, -1)
        }, () => {
            const currExp = this.state.expression;
            this.setState({
                prevChar: currExp.split('').pop()
            })
        })
    }

    calcStringExpression = expArr => {
        let [a, operator = '+', b = 0, ...rest] = expArr;
        b = isNaN(b) ? 0 : b;
        const calcVal = this.arthimeticOperations(operator)(a, b);
        return [calcVal, ...rest]
    }

    calculateResult = (expression) => {
        let result = [];
        const copy = expression;

        expression = expression.replace(/[0-9]+/g, "#");
        const numbers = copy.split(/[^0-9.]+/);
        const operators = expression.split("#").filter(n => n);

        for(let i = 0; i < numbers.length; i++){
             result.push(parseInt(numbers[i]));
             if (i < operators.length) result.push(operators[i]);
        }
        let resultLen = result.length;
        while(resultLen > 1) {
            let newCalcResult = this.calcStringExpression(result);
            resultLen = this.calcStringExpression(newCalcResult).length;
            result = newCalcResult;
        }
        const finalResult = this.calcStringExpression(result)[0];
        this.setState({
            result: finalResult,
            expression: ''
        })
    }

    arthimeticOperations(type) {
        const arthOps = {
            '+': (a, b) => a + b,
            '-': (a, b) => a - b,
            '*': (a, b) => a * b,
            '/': (a, b) => a / b,
        }
        return arthOps[type]
    }

    render() {
        return (
            <div className="calculator">
                <Result expression={this.state.expression} result={this.state.result} />
                <Keypad keyClicked={(e) => this.updateExpression(e) }/>
            </div>
        )
    }
}
