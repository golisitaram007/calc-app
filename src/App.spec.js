import React from 'react';
import renderer from 'react-test-renderer';
import Calculator from './components/Calculator';

describe('Calculator', () => {
    test('snapshot renders', () => {
      const component = renderer.create(<Calculator />);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
});

describe('My Test Suite', () => {
    it('My Test Case', () => {
        expect(true).toEqual(true);
    });
});