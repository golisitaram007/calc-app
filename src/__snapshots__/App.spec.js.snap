// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator snapshot renders 1`] = `
<div
  className="calculator"
>
  <div
    className="calcResult"
  >
    <div
      className="expression"
    >
      
    </div>
    <div
      className="finalResult"
    >
      
    </div>
  </div>
  <div
    className="calcKeypad"
  >
    <div
      className="keyRow row1"
    >
      <button
        className="clearBtn"
        data-key="clear"
        onClick={[Function]}
      >
        Clear
      </button>
      <button
        className="actionBtn bkSpBtn"
        data-key="Backspace"
        onClick={[Function]}
      >
        ←
      </button>
      <button
        className="divisionBtn actionBtn"
        data-key="/"
        onClick={[Function]}
      >
        ÷
      </button>
    </div>
    <div
      className="keyRow row2"
    >
      <button
        data-key="7"
        onClick={[Function]}
      >
        7
      </button>
      <button
        data-key="8"
        onClick={[Function]}
      >
        8
      </button>
      <button
        data-key="9"
        onClick={[Function]}
      >
        9
      </button>
      <button
        className="actionBtn"
        data-key="*"
        onClick={[Function]}
      >
        ×
      </button>
    </div>
    <div
      className="keyRow row3"
    >
      <button
        data-key="4"
        onClick={[Function]}
      >
        4
      </button>
      <button
        data-key="5"
        onClick={[Function]}
      >
        5
      </button>
      <button
        data-key="6"
        onClick={[Function]}
      >
        6
      </button>
      <button
        className="actionBtn"
        data-key="-"
        onClick={[Function]}
      >
        -
      </button>
    </div>
    <div
      className="keyRow row4"
    >
      <button
        data-key="1"
        onClick={[Function]}
      >
        1
      </button>
      <button
        data-key="2"
        onClick={[Function]}
      >
        2
      </button>
      <button
        data-key="3"
        onClick={[Function]}
      >
        3
      </button>
      <button
        className="actionBtn"
        data-key="+"
        onClick={[Function]}
      >
        +
      </button>
    </div>
    <div
      className="keyRow row5"
    >
      <button
        className="zeroBtn"
        data-key="0"
        onClick={[Function]}
      >
        0
      </button>
      <button
        className="equalBtn actionBtn"
        data-key="Enter"
        onClick={[Function]}
      >
        =
      </button>
    </div>
  </div>
</div>
`;
